dcos_launch/terraform.py

```diff
@@ -0,0 +1,283 @@
+import json
+import logging
+import os
+import re
+import shutil
+import subprocess
+import uuid
+import zipfile
+
+import requests
+from cryptography.hazmat.primitives import serialization
+
+import dcos_launch.config
+import yaml
+from dcos_launch import gcp, util
+from dcos_launch.platforms import aws
+from dcos_test_utils import helpers
+
+log = logging.getLogger(__name__)
+IP_REGEX = '(\d{1,3}.){3}\d{1,3}'
+IP_LIST_REGEX = '\[[^\]]+\]'
+
+
+def _get_ips(prefix: str, info: str) -> list:
+    ips = []
+    prefix += ' = '
+    m = re.search('{}{}'.format(prefix, IP_REGEX), info)
+    if m:
+        ips = [m.group(0)[len(prefix):]]
+    else:
+        m = re.search('{}{}'.format(prefix, IP_LIST_REGEX), info)
+        if m:
+            # remove prefix
+            s = m.group(0)[len(prefix):]
+            # remove whitespace
+            s = "".join(s.split())
+            # remove brackets
+            s = s[1:-1]
+            ips = s.split(',')
+    return ips
+
+
+def _convert_to_describe_format(ips: list) -> list:
+    # TODO private ips not currently outputted in terraform-dcos
+    return [{'private_ip': None, 'public_ip': ip} for ip in ips]
+
+
+class TerraformLauncher(util.AbstractLauncher):
+    def __init__(self, config: dict, env=None):
+        if env:
+            os.environ.update(env)
+        self.config = config
+        self.init_dir = dcos_launch.config.expand_path('', self.config['init_dir'])
+        self.cluster_profile_path = os.path.join(self.init_dir, 'desired_cluster_profile.tfvars')
+        self.dcos_launch_root_dir = os.path.abspath(os.path.join(self.init_dir, '..'))
+        self.terraform_binary = os.path.join(self.dcos_launch_root_dir, 'terraform')
+        self.default_priv_key_path = os.path.join(self.init_dir, 'key.pem')
+
+    def terraform_cmd(self):
+        """ Returns the right Terraform invocation command depending on whether it was installed by the user or by
+        dcos-launch.
+        """
+        binary = self.terraform_binary
+        if not os.path.exists(binary):
+            binary = 'terraform'
+        return binary
+
+    def create(self):
```

Comment:
Is this the only place the existence of `terraform` is checked? Can this be moved to `terraform_cmd` instead? There is no guarantee that a user will go through the standard flow.

Reply:
If the user tries to describe or delete a cluster he has never created, why would I want to check terraform is installed? The CLI will throw him an error because he doesn't have a cluster_info.json

Reply:
@colin-msphere Interesting point to bring up. This would be valid for every other dcos-launch provider EXCEPT Terraform. The reason for this is you need more than a cluster_info.json to interact with a dcos-launch-terraform cluster. When create() is called, both terraform init and terraform apply are called inside the terraform directory we created for this cluster. Those terraform commands will generate a .terraform directory with data essential to interact with that cluster.